Tape expansion reserves value of the is_sampled attribute

antalszava opened the pull request:
**Context:**

Devices that do not support an operation try to apply its decomposition. In tape mode this triggers tape expansion. However, in such cases, new tapes are being created and information for the `is_sampled` attribute is lost.

This causes problems when `analytic=True` is set under the hood and the QNode returns samples:
```python
import pennylane as qml

qml.enable_tape()

dev = qml.device('default.qubit', wires=1, analytic=True)

dev.operations.remove('T')

@qml.qnode(dev)
def circuit():
    qml.T(wires=0)
    return qml.sample(qml.PauliZ(0))

circuit()
```

```
~/pennylane/_qubit_device.py in sample(self, observable)
    681         if isinstance(name, str) and name in {"PauliX", "PauliY", "PauliZ", "Hadamard"}:
    682             # Process samples for observables with eigenvalues {1, -1}
--> 683             return 1 - 2 * self._samples[:, device_wires[0]]
    684 
    685         # Replace the basis state in the computational basis with the correct eigenvalue.

TypeError: 'NoneType' object is not subscriptable
```

**Description of the Change:**

* Explicitly sets the `is_sampled` attribute after tape expansion.

**Benefits:**

* Sampling works fine for operations that are not supported.
* Usually, the `CircuitGraph` object can be queried to get this information. However, as the circuit graph is not necessarily constructed after circuit expansion, it is less costly to simply iterate over the observables without constructing the circuit graph.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Demo using the Cirq plugin.

thisac commented on pennylane/tape/qnode.py:
```diff
@@ -447,6 +447,9 @@ def construct(self, args, kwargs):
                 stop_at=lambda obj: self.device.supports_operation(obj.name),
             )
 
+            # Check the observables without needing to create the circuit graph
+            self.qtape.is_sampled = any(obs.return_type == Sample for obs in self.qtape.observables)
```
Really nice catch @antalszava! :100: 

I'm just wondering why the `is_sampled` method was changed to an attribute. Maybe this is nicer (and a bit more effective) to have it this way rather than always checking for `Sample` return types in the observables. :thinking:

antalszava replied:
Hmm, so `is_sampled` is an attribute of `QuantumTape`, but a property of the `CircuitGraph`. Here at times, the circuit graph was not constructed, so simply checking the observables directly avoids the overhead of constructing the circuit graph

thisac replied:
I mean, an option would be to have `QuantumTape.is_sampled` a property instead of an attribute, as is the case for `CircuitGraph`. No need to create the graph in either case; it would be a small change. I'm not really advocating for either though, just a bit curious about the reasoning. :slightly_smiling_face:

josh146 replied:
> Hmm, so is_sampled is an attribute of QuantumTape, but a property of the CircuitGraph. 

This was likely my fault, I didn't check that it was a property on the graph and just made it an attribute on the tape. However,

* this is probably not too much of an issue, since the tape is designed to be (mostly) non-mutable, whereas the circuit graph was very mutable.

* It was mostly a temporary fix, I needed the tape to 'look' like a CircuitGraph from a duck-typing perspective, so that existing devices would accept it. With the thinking that soon we would be able to drop this backward compatibility :)

thisac commented on tests/tape/tapes/test_tape.py:
```diff
@@ -840,6 +840,26 @@ def test_expand_tape_multiple_wires_non_commuting(self, ret):
         with pytest.raises(qml.QuantumFunctionError, match="Only observables that are qubit-wise"):
             tape.expand(expand_measurements=True)
 
+    def test_is_sampled_reserved_after_expansion(self, monkeypatch, mocker):
```
(thinking aloud) I think this test looks good, but I'm contemplating whether this is too broad of a test (not unit-test-like enough), but I'm not sure how to do it otherwise.

antalszava replied:
Yeah, think that's a fair point. One thing I found challenging with creating more of a unit test is that a tape is created and expanded in `QNode.construct` along with other updates (perhaps later down the line we could think about separating these into different methods).

thisac replied:
Without being to familiar with writing tests, etc. I would say a good option could simply be to be better at separating the unit tests and the integration tests (and this could perhaps be part of the latter, right?). Then no unit test would be needed for this case at all.

antalszava replied:
Hmm, it seems that integration tests were not too separated so far for tapes. Might just leave it here for now and we could move it if we start the separation into a new file.

thisac commented on pennylane/tape/operation.py:
```diff
@@ -63,6 +63,8 @@ def operation_expand(self):
         operations decomposition, or if not implemented, simply
         the operation itself.
     """
+    # TODO: rename the following usage of JacobianTape to better describe that
+    # it's used for operation expansion
```
Do you mean simply renaming it to e.g. `op_expansion_tape`? I'm not against it per se, but I'm thinking it's quite clear already from the method + docstring, and it's only restricted to this quite small method. Maybe this is just something that can be done in this PR (since it's a very small, cosmetic change only) in that case instead of leaving it as a future TODO? :slightly_smiling_face:

antalszava replied:
The comment more so implies that we are creating a `JacobianTape` which is not necessarily used for jacobian computations. Perhaps we could a) rename  `JacobianTape` or b) create a new type of tape that is more specific to tape expansions.

thisac replied:
Oh, I see. In that case, this is a bit more complex. I though you meant renaming the name `tape` and not the tape `Jacobiantape`. :laughing:

josh146 replied:
Oh good catch @antalszava! I just had a look at the source code for `QuantumTape.expand()`, and it looks like this doesn't need to be a `JacobianTape`, you should be able to replace it with a simple `QuantumTape` with no side effect.

thisac reviewed (approved):
Really nice fix @antalszava. :100: Now the demo works with tape mode!

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -371,6 +371,10 @@
 <h3>Bug fixes</h3>
 
+* Fixes an issue with tape expansions where information about sampling
+  (``is_sampled`` tape attribute) was lost previously.
```
```suggestion
  (specifically the `is_sampled` tape attribute) was not preserved.
```

josh146 reviewed (approved):
Thanks Antal for finding this! 💯 

Happy for it to be merged as is. One (minor) question though: would it be better to have this at the end of tape expansion, rather than in the QNode? E.g., around here?

https://github.com/PennyLaneAI/pennylane/blob/2ae02637e60319be74f164286cc6991451ee886c/pennylane/tape/tapes/tape.py#L150

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -371,6 +371,10 @@
 
+* Fixes an issue with tape expansions where information about sampling
+  (``is_sampled`` tape attribute) was lost previously.
+  [(#956)](https://github.com/PennyLaneAI/pennylane/pull/1027)
```
```suggestion
  [(#1027)](https://github.com/PennyLaneAI/pennylane/pull/1027)
```

antalszava commented:
Thanks so much for the suggestions!

1. Changed to have `QuantumTape` in the expansion logic.
2. Moved the update of `is_sampled` to the end of the expansion (instead of the QNode).
